### PR TITLE
Support multiple assignees, completers and selected_options

### PR DESCRIPTION
**Pull request by keithtom**
New requirements moved us from tasks complete by a single assignee to multiple assignees and completers.

image

Designs
https://www.figma.com/file/LgzEVh8jzborTam1Uz07e4/SSJ---Mockups?node-id=1193-11331&t=rYfKJkTzHpLj75Wz-0

### CODE REVIEW

File: app/controllers/v1/workflow/processes_controller.rb [hunk 1]
```diff
@@ -24,13 +24,6 @@ def index
 
     options = {include: ['workflow', 'steps', 'steps.documents']}
 
-    if params[:self_assigned]
-      processes = processes.select do |process|
-        process.steps.where(assignee_id: current_user.person_id, completed: false).count > 0
-      end
-      options[:params] = { assignee_id: current_user.person_id, self_assigned: true }
-    end
-
```
Comment (keithtom) [inline app/controllers/v1/workflow/processes_controller.rb:33]: no longer used feature (was for todo list/assignment on dashboard)

File: app/models/person/ssj.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  # Person logic specifically related to SSJ
+  module SSJ
+    extend ActiveSupport::Concern # https://dev.to/software_writer/how-rails-concerns-work-and-how-to-use-them-gi6
+
+    included do
+      has_one :ssj_team_member, class_name: "SSJ::TeamMember", foreign_key: 'person_id'
+      has_one :ssj_team, through: :ssj_team_member
+    end
+  end
+end
```
Comment (keithtom) [inline app/models/person/ssj.rb:11]: Minor cleanup, just introducing modules for storing person logic related to large subcategories like SSJ and workflow

File: app/serializers/v1/workflow/step_assignment_serializer.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+class V1::Workflow::StepAssignmentSerializer < ApplicationSerializer
+  set_id :created_at # exception to the rule, we don't care about sharing this
```
Comment (styliii) [inline app/serializers/v1/workflow/step_assignment_serializer.rb:2]: Why are we setting the id to `created_at` timestamp? Shouldn't we just use the id?

File: app/services/workflow/instance/step/assign_person.rb [hunk 4]
```diff
@@ -7,8 +7,10 @@ def initialize(step, person)
       end
 
       def run
-        @step.assignee_id = @person.id
-        @step.save!
+        # NOTE: if the step worktype was "only 1 assigner", we'd unassign first.
+        @step.assignments.find_or_create_by!(assignee: @person)
+        @step.assigned = true
+        @step.save
```
Comment (styliii) [inline app/services/workflow/instance/step/assign_person.rb:13]: Should we check if we should update the step's process' status to `in_progress`?

File: app/services/workflow/instance/process/complete.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      # after we complete, we need to see what's unlocked.
+      # that's done later since state is changed..
+      def initialize(process)
+        @process = process
+      end
+
+      def run
+        # @process.completed = true # do processes have these fields?  separate this to another table?
+        # @process.completed_at = DateTime.now # this is on the step.
+        @process.status = V1::Statusable::DONE
```
Comment (styliii) [inline app/services/workflow/instance/process/complete.rb:17]: 🤔 I don't see a `status` column on `process`. Only `completion_status`, which is an enum.

File: app/services/workflow/instance/step/uncomplete.rb [hunk 6]
```diff
@@ -1,14 +1,50 @@
...
+      def check_relock_step_dependencies!
+      end
+
+      def update_process_completed_counter_cache
+      end
+
+      def update_process_completion_status
+      end
+
+      def unstart_process
```
Comment (styliii) [inline app/services/workflow/instance/step/uncomplete.rb:47]: Don't forget to fill these out!

also, is there a reason why you moved them out of the model after save hooks? I'm assuming you wanted it to be more explicit when the status updates are being changed.

File: app/models/workflow/instance/step.rb [hunk 7]
```diff
@@ -2,18 +2,21 @@ module Workflow
...
+    has_many :documents, as: :documentable
+    
     before_create :set_position
-    after_save :update_process
-    after_destroy :update_process
-
-    DEFAULT_INCREMENT = 1000
 
+    scope :complete, -> { where(completed: true) }
+    scope :incomplete, -> { where(completed: [false, nil]) }
```
Comment (styliii) [inline app/models/workflow/instance/step.rb:16]: nit: `s/complete/completed` and `s/incomplete/incompleted`

File: app/services/workflow/instance/step/complete.rb [hunk 8]
```diff
@@ -7,27 +7,92 @@ class Complete < BaseService
...
+          @process.save!
+        end
+      end
+
+      # check if the that was the last step and complete the process
+      # TODO: move this to a Process::Complete service
+      def complete_process
+        if @process.completed_steps_count == @process.steps_count
+          @process.completed_at = DateTime.now
+          @process.save!
```
Comment (styliii) [inline app/services/workflow/instance/step/complete.rb:84]: Need to call @process.done! here.

File: app/services/workflow/instance/step/assign_person.rb [hunk 9]
```diff
@@ -7,8 +7,10 @@ def initialize(step, person)
       end
 
       def run
-        @step.assignee_id = @person.id
-        @step.save!
+        # NOTE: if the step worktype was "only 1 assigner", we'd unassign first.
```
Comment (styliii) [inline app/services/workflow/instance/step/assign_person.rb:10]: Should we call start process here?

File: app/controllers/v1/workflow/steps_controller.rb [hunk 10]
```diff
@@ -1,37 +1,31 @@
 class V1::Workflow::StepsController < ApiController
+  before_action :find_step, except: [:create]
```
Comment (styliii) [inline app/controllers/v1/workflow/steps_controller.rb:2]: Nice! Cleans everything up.

File: spec/services/workflow/instance/step/unassign_person_spec.rb [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  context "when there are assigned people" do
+    before do
+      Workflow::Instance::Step::AssignPerson.run(step, person)
+      Workflow::Instance::Step::AssignPerson.run(step, person2)
+    end
+
+    it "unassigns a person from a step" do
+      expect(step.assignments.count).to eq(2)
+      expect(step.assignments).to include(
+        have_attributes(assignee: person)
```
Comment (styliii) [inline spec/services/workflow/instance/step/unassign_person_spec.rb:17]: TIL. Never knew you could do this rspec syntax.

**Pull request merged**